#1042 Add automatic backup option

marstaj opened the pull request:
#1042 Add automatic backup option

marstaj committed:
feat: Add backup export schedule functionality

This commit introduces the ability to schedule automatic backup exports.

Key changes:
- Added UI elements for selecting backup export schedule options (e.g., Every Hour, Every Day, Every Week).
- Implemented `BackupExportScheduleWorker` to handle automatic backup creation based on the selected schedule.
- Introduced use cases for reading and writing backup JSON to files (`ReadBackupJsonFromFileUseCase`, `WriteBackupJsonToFileUseCase`).
- Added a use case for creating `BackupScheme` from JSON (`CreateBackupSchemeFromJsonUseCase`).
- Added a use case for creating backup JSON string (`CreateBackupJsonUseCase`).
- Implemented logic in `BackupExportViewModel` to manage export scheduling and one-off exports.
- Updated `MainViewModel` and `MainActivity` to refresh the backup export schedule on app start.
- Added `BackupFileName` object to generate standardized backup file names.
- Updated string resources for new schedule options and messages.
- Added necessary WorkManager and Hilt dependencies.

marstaj committed:
feat: Enhance automatic backup by limiting stored files

This commit introduces a mechanism to limit the number of automatic backup files stored. The system will now:
- Retain only the 5 most recent backup files.
- Automatically delete older backup files to manage storage space.
- Backup files are identified by a specific prefix and file type (`showly_export_*.json`) and sorted by the timestamp embedded in their filenames for deletion priority.

Additionally, constants related to backup file naming (prefix, file type, date/time pattern, and MIME type) have been centralized in `BackupFileName.kt`.

marstaj committed:
Fix: Enable initial delay for backup export worker

The initial delay for the `BackupExportScheduleWorker` was previously
commented out. This commit re-enables the initial delay, ensuring
that the periodic work request respects the scheduled duration and unit
before its first execution.

marstaj committed:
Fix: Remove the 15 minute backup option

michaldrabik commented:
@marstaj One thing I noticed now is that folder selection is popping up even if you turn the schedule off.

michaldrabik reviewed (changes requested):
One thing I noticed now is that folder selection is popping up even if you turn the schedule off.

marstaj commented:
Good point, I'll fix that.

marstaj committed:
Merge branch 'master' into feature/1042_Add_automatic_backup_option

marstaj committed:
Refactor: Improve backup export schedule handling

This commit refactors the backup export scheduling logic:

- Directly saves "OFF" schedule without requiring folder selection.
- Shows confirmation snackbar after successfully selecting a schedule and folder.
- Introduces `saveExportBackupScheduleOff()` in `BackupExportViewModel` for explicitly setting the schedule to OFF.

marstaj commented:
I've made the change to only ask for folder picker when the schedule is not OFF. Also moved around the snackbar message a little as I noticed that the message was not showing with the folder picker correctly.

michaldrabik reviewed (commented):
Overall thanks for great job. Commented on 1 thing and it might be ready to ship. @marstaj

marstaj commented on ui-backup/src/main/java/com/michaldrabik/ui_backup/features/export/workers/BackupExportScheduleWorker.kt:
```diff
@@ -0,0 +1,224 @@
+    }
+
+    if (backupFiles.size > MAX_BACKUPS) {
+      backupFiles.sortBy { it.second } // Sort by name, which is effectively by date
```
Yes, that's fair. I'll fix.

marstaj committed:
Refactor: Sort backups by last modified timestamp for deletion

This commit updates the backup deletion logic in `BackupExportScheduleWorker`.
Instead of relying on the filename (which includes a timestamp) for sorting, backups are now sorted by their actual last modified timestamp. This ensures more accurate deletion of the oldest backups when exceeding `MAX_BACKUPS`.

- Modified `deleteOldBackups` to query `DocumentsContract.Document.COLUMN_LAST_MODIFIED`.
- Changed `backupFiles` list to store a `Triple` including the last modified timestamp.
- Updated sorting logic to use the last modified timestamp.

marstaj committed:
Merge branch 'master' into feature/1042_Add_automatic_backup_option

marstaj commented:
@michaldrabik sorry for the delay, I fixed  this when I read your comment but then tested for a few days and forgot that I never pushed the code 🤦‍♂️. Hopefully everything is in order now.
